tools/worker-runner/registration/registration.go

```diff
@@ -96,22 +98,81 @@ func (reg *RegistrationManager) WorkerStarted() error {
 
 	untilExpire := time.Until(reg.state.CredentialsExpire)
 	reg.credsExpireTimer = time.AfterFunc(untilExpire-30*time.Second, func() {
-		if reg.proto != nil && reg.proto.Capable("graceful-termination") {
-			log.Println("Taskcluster Credentials are expiring in 30s; stopping worker")
-			reg.proto.Send(workerproto.Message{
-				Type: "graceful-termination",
-				Properties: map[string]interface{}{
-					// credentials are expiring, so no time to shut down..
-					"finish-tasks": false,
-				},
-			})
+		// Prefer to update the worker's credentials, but..
+		if reg.proto.Capable("new-credentials") {
+			reg.reregisterWorker()
+			// fall back to just shutting down the worker
+		} else if reg.proto.Capable("graceful-termination") {
+			reg.terminateWorker()
+		} else {
+			panic("credentials expiring, but no way to tell the worker")
 		}
 	})
 
 	return nil
 }
 
+// Request new credentials and send them to the worker, falling back to requesting
+// a worker termination if anything fails.  Called with reg.state locked
+func (reg *RegistrationManager) reregisterWorker() {
+	log.Println("Taskcluster Credentials are expiring in 30s; re-registering")
```

Comment:
I wonder if we should give a bit more headroom than 30s in the case of retries, etc.

Reply:
Good point -- and we want to leave some leeway for restarts, etc. too.  I'll make a more sophisticated check.